crates/holochain_sqlite/tests/db_wrapper.rs

```diff
@@ -0,0 +1,383 @@
+use holochain_sqlite::db::{num_read_threads, PermittedConn, WriteManager};
+use holochain_sqlite::error::DatabaseError;
+use holochain_sqlite::error::DatabaseError::{DbConnectionPoolError, Timeout};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::Arc;
+
+#[cfg(feature = "test_utils")]
+mod common;
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[test]
+fn get_connections_from_pool() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let number_of_connections_received = Arc::new(AtomicUsize::new(0));
+    let mut join_handles = Vec::new();
+
+    // Number of competing jobs
+    for _ in 0..200 {
+        let my_handle = db_handle.clone();
+        let my_number_of_connections_received = number_of_connections_received.clone();
+        join_handles.push(std::thread::spawn(move || {
+            // Number of database connections opened over time
+            for _ in 0..100 {
+                let _ = my_handle.conn().unwrap();
+                my_number_of_connections_received.fetch_add(1, Ordering::SeqCst);
+
+                std::thread::sleep(std::time::Duration::from_millis(100));
+            }
+        }));
+    }
+
+    for jh in join_handles.into_iter() {
+        jh.join().unwrap();
+    }
+
+    assert_eq!(
+        20_000,
+        number_of_connections_received.load(Ordering::Acquire)
+    )
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[test]
+fn pool_size_is_limited() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    // Assumes knowledge of the implementation, readers plus one writer
+    let pool_size = num_read_threads() + 1;
+
+    let mut hold_connections = Vec::with_capacity(pool_size);
+    for _ in 0..pool_size {
+        let c = db_handle.conn().unwrap();
+        hold_connections.push(c)
+    }
+
+    let mut failed_count = 0;
+    for _ in 0..3 {
+        // Cannot get any more connections on this thread
+        match db_handle.conn() {
+            // This is caused by a timeout at the connection pool level but could be more specific
+            Err(DbConnectionPoolError(_)) => {
+                // Could not get a connection, this is the desired outcome. Sleep and try again
+                failed_count += 1;
+                std::thread::sleep(std::time::Duration::from_millis(10));
+            }
+            Err(e) => {
+                panic!("Got an unexpected error - {:?}", e);
+            }
+            Ok(_c) => {
+                panic!("Should not have been able to get a connection");
+            }
+        }
+    }
+
+    assert_eq!(3, failed_count);
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
```

Comment:
@ThetaSinner - none of the tests in here are using `#[tokio::test(flavor = "multi_thread")]` - and we exclusively use the multi_thread runtime. Do you know, did they change the default? Even if they did, perhaps stating the flavor explicitly would make it clearer we are using the same environment in testing and non-testing code.

Reply:
Good point, I've updated these to use `#[tokio::test(flavor = "multi_thread")]`